src/coreclr/jit/gentree.cpp

```diff
@@ -21848,6 +21858,10 @@ GenTree* Compiler::gtNewSimdCreateBroadcastNode(var_types   type,
             {
                 uint32_t cnsVal = static_cast<uint32_t>(op1->AsIntConCommon()->IntegralValue());
 
+#if defined(TARGET_XARCH)
+                vecCon->SetCreatedFromScalar();
```

Comment:
I wonder if we need to track the size of the scalar it was created from as this is going to miss some cases...

That is, a constant broadcast from an `8-bit`, `16-bit`, or `32-bit` scalar can be used as part of `32-bit` or `64-bit` constant since every `32-bit` or `64-bit` element will be identical. While one created from a `64-bit` scalar can only potentially be used as part of a `32-bit` constant (if the upper/lower 32-bits are themselves identical).

Alternatively, we could avoid the flag altogether. Since lowering may need to check such cases anyways (to handle such scenarios) we could just have it lower to a `BroadcastScalarToVector` node and contain that instead.

Thoughts?

Reply:
That make sense to me, and I am trying to make the methodology more clear.
So we want, when the use of broadcast node is an embedded broadcast compatible node, instead of lowering the `Broadcast(Const)` into a constant vector, we simply keep the `Broadcast -> Const` form, lower them, and make the broadcast node contained by the embedded broadcast compatible node.
Does that match what we are thinking?

Reply:
if we follow this, we may face some difficulty since the `broadcast->const` form is converted to constant vector when importation, in [impSpecialIntrinsic()](https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/hwintrinsicxarch.cpp#L1097) by method [gtNewSimdCreateBroadcastNode()](https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/gentree.cpp#L21756). It might be late to catch the `broadcast->const` form when lowering. 

On the other hand, in [LowerHWIntrinsicCreate](https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/lowerxarch.cpp#L2106), here we have some similar `constant` check logics similar to that in import, I wonder if [this branch](https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/lowerxarch.cpp#L2111) will be hit or not.

Reply:
> Does that match what we are thinking?

Yes. 

> if we follow this, we may face some difficulty since the broadcast->const form is converted to constant vector when importation,

This is fine and generally desirable so that constant folding can still occur.

We would just need to have an additional step in lowering that observes `GenTreeVecCon` and its user. If its user is `EmbBroadcastCompatible` then we would get the input size of the user (32-bit/64-bit) and check if all dwords/qwords are identical. If they are, we can convert the constant into a `broadcast` node and the containment support for broadcast will light up at that point.

We would similarly get the user and avoid converting `Broadcast` into a `GenTreeVecCon` if the user is `EmbBroadcastCompatible`.

Reply:
That sounds reasonable to me.